python/pyspark/worker.py

```diff
@@ -841,6 +841,27 @@ def _remove_partition_by_exprs(self, arg: Any) -> Any:
             "the query again."
         )
 
+    # Compute the set of UDTF result columns whose types are not nullable.
+    # We will check that the UDTF does not return None values for these columns below.
+    non_nullable_result_cols = set()
+    for i, field in enumerate(return_type):
```

Comment:
I'm afraid this is not enough as potentially this may also happen in the nested fields, like array, map, and struct types.

Reply:
Also we should check the other type of UDFs, including Scala UDFs.

Reply:
Thanks for your reviews! I updated this PR to also check recursively for `None` values within array, struct, and map values as well.
We can certainly consider other types of UDFs as well later (including Scala UDFs); it seems possible to decouple that work from Python UDTFs here, so I'll leave this PR to focus on the latter for now.